### PR TITLE
FRONTEND-4243 :: Feature :: Harmony-typescript NetworkDataSource

### PR DESCRIPTION
**Pull request by urijp**
**Asana task link:**
https://app.asana.com/0/1109863238977521/1202775968264243/f

**Pull request information:**
This PR contains de implementation of the generic network data source. The relevant changes here are:

- NetworkDataSource implementation
- NetworkQuery implementation
- HttpRequestBuilder refactor, removing the mapping within this class to always use DataSourceMapper
- Creation of two helpers to provide DefaultNetworkDataSourceMapper and DefaultNetworkDataSourceArrayMapper.

### CODE REVIEW

File: packages/core/src/repository/data-source/network.data-source.ts
Comment (urijp) [inline packages/core/src/repository/data-source/network.data-source.ts:59]: value will always have priority over query when this is defined.

File: packages/core/src/repository/query/network-query.ts
Comment (urijp) [inline packages/core/src/repository/query/network-query.ts:20]: The http request methods are not added since the query class have it implicit. Actually, the actions would have it implicitely included but we need Post and Put network queries to distinguis these methods in the Put method.

File: packages/core/src/repository/data-source/network.data-source.ts
Comment (urijp) [inline packages/core/src/repository/data-source/network.data-source.ts:57]: Open to discussion:

The idea of sharing this method on all types of requests comes from the http specifications, when we can have query parameters + url parameters + body in GET/POST/PUT/DELETE methods, so we have the most generic network datasource we can have.

Our agreement with the backend team is to never use URL parameters, and never use body in GET/DELETE methods. We could change this datasource to match that, but this will only work against our own backends.

File: packages/core/src/repository/data-source/network.data-source.ts
Comment (urijp) [inline packages/core/src/repository/data-source/network.data-source.ts:18]: Since we are now mapping in the DataSourceMapper, NetworkDataSource will always return an unknown type of object.

File: packages/core/src/repository/query/network-query.ts
Comment (Alex-DA) [inline packages/core/src/repository/query/network-query.ts:4]: @orioljp Question: the endpoint includes the domain or only the path?.

Comment (lucianosantana) [reply packages/core/src/repository/query/network-query.ts:4]: It should be only the path. See packages/core/src/repository/data-source/api-request.service.ts , builder method. The query.endpoint is being used when calling the method. We could change the parameter's name if it makes it clearer.

**Comment by lucianosantana** [comment]
@orioljp why is it still flagged as WIP ? What is missing?

File: packages/core/src/repository/query/network-query.ts
Comment (Alex-DA) [reply packages/core/src/repository/query/network-query.ts:4]: @orioljp @lucianosantana and this behaviour allow us to point to two APIs with different domains?.

**Review by Alex-DA — APPROVED** [review]

File: packages/core/src/repository/query/network-query.ts
Comment (lucianosantana) [reply packages/core/src/repository/query/network-query.ts:4]: Good point @Alex-DA . The base URL is defined on the `ApiRequestService`. In this case, we'd have to create a new service instance and a new NetworkDataSource for each API base URL. So the setting is done via dependency injection. We could change it to have the base URL defined on every query. Both are a bit of extra work. Having it on the query seems to be more work for me.

Comment (orioljp) [reply packages/core/src/repository/query/network-query.ts:4]: @Alex-DA Yes, we can have many base urls but not in the same datasource, since a datasource is one, two base urls would mean two sources of data, and this is managed by providing 2 different ApiRequestServices with different urls. Anyway, the behaviour of the baseUrl managing has not changed in this PR

**Comment by urijp** [comment]
@lucianosantana yes, it's in progress because IMO we need an example or proper testing and this is not done mainly because I want to validate the approach with all of you before facing it. Could you check it @lucianosantana @doup @joselufo ?

**Comment by lucianosantana** [comment]
> @lucianosantana yes, it's in progress because IMO we need an example or proper testing and this is not done mainly because I want to validate the approach with all of you before facing it. Could you check it @lucianosantana @doup @joselufo ?

The approach seems correct to me.

**Comment by joselufo** [comment]
I had a call with @orioljp, and I think the approach is correct. It has some differences with the Kotlin and Swift implementation, but its development makes sense.

**Review by joselufo — APPROVED** [review]

File: packages/core/src/repository/data-source/network.data-source.ts
Comment (doup) [inline packages/core/src/repository/data-source/network.data-source.ts:30]: **Nit:** isn't the convention to put the `$` at the end? So it reads like "requests"; since it's a stream, multiple values might be emitted. Although we know only one will be emitted.

See: https://angular.io/guide/rx-library#naming-conventions-for-observables

File: packages/core/src/repository/data-source/network.data-source.ts
Comment (doup) [inline packages/core/src/repository/data-source/network.data-source.ts:23]: **Nit:** I don't think this is needed, this kind of generic is usually inferred from the context. **You can do the same in all the cases**. I've tried to remove it… at the compiler is OK with it.

I'm not sure 100% why it works though. I think it's because we've `unknown` as the return type.

```suggestion
            return lastValueFrom(this.getRequestWithParameters(query).get());
```

File: packages/core/src/repository/data-source/network.data-source.ts
Comment (doup) [reply packages/core/src/repository/data-source/network.data-source.ts:59]: `value` is not being passed to `getRequestWithParameters` in the `put` method. So, this will never run. :P

File: packages/core/src/repository/data-source/network.data-source.ts
Comment (doup) [reply packages/core/src/repository/data-source/network.data-source.ts:57]: I would leave it as it is, as it's more flexible.

File: packages/core/src/repository/data-source/network.data-source.ts
Comment (doup) [inline packages/core/src/repository/data-source/network.data-source.ts:84]: Won't `VoidMapper` error in case we use a `value` in the `put` method? So, it should be `BlankMapper`? (I think that's our "noop" mapper).

File: packages/core/src/repository/data-source/network.data-source.ts
Comment (doup) [inline packages/core/src/repository/data-source/network.data-source.ts:83]: **Nit:** I think you can skip the explicit types.

```suggestion
        new JsonDeserializerMapper(type),
```

File: packages/core/src/repository/data-source/network.data-source.ts
Comment (doup) [inline packages/core/src/repository/data-source/network.data-source.ts:79]: **Nit:** I think you can skip the explicit types.

```suggestion
    return new DataSourceMapper(
```

File: packages/core/src/repository/data-source/network.data-source.ts
Comment (doup) [inline packages/core/src/repository/data-source/network.data-source.ts:98]: **Nit:** I think you can skip the explicit types.

```suggestion
    return new DataSourceMapper(
        dataSource,
        dataSource,
        dataSource,
        new ArrayMapper(
            new JsonDeserializerMapper(type),
```

File: packages/core/src/repository/data-source/network.data-source.ts
Comment (doup) [inline packages/core/src/repository/data-source/network.data-source.ts:100]: Same as in `provideDefaultNetworkDataSource`?

File: packages/core/src/repository/query/network-query.ts
Comment (doup) [reply packages/core/src/repository/query/network-query.ts:4]: Maybe we can change this from `endpoint` to `path`? Or, we can add a docblock for the `endpoint` var explaining that the `baseUrl` is managed elsewhere.

Comment (doup) [inline packages/core/src/repository/query/network-query.ts:4]:
```suggestion
    constructor(
        /** Endpoint path without `baseUrl`. Base URL is managed by `ApiRequestService`. */
        public readonly endpoint: string,
    ) {}
```

File: packages/core/src/repository/query/network-query.ts
Comment (doup) [inline packages/core/src/repository/query/network-query.ts:6]: See comment regarding `AngularHttpRequestBuilder.body`; we could encode "no body" as `undefined` instead of using an empty string?

File: packages/core/src/repository/query/network-query.ts
Comment (doup) [inline packages/core/src/repository/query/network-query.ts:3]: I find weird the `NetworkQuery` class setup:
- is `abstract`, but it doesn't expose any abstract method. Remove `abstract`?
- is exported; but it's useless on it's own, as we wouldn't know to which HTTP method to map in the network data-source. So, either:
    - We don't export this `NetworkQuery`.
    - Or, we export it, but we add a public `method` property. In which case the base class will have an "open" `GET | POST | PUT | DELETE` union, while the subclasses already specify their specific method, e.g. `GetNetworkQuery.method = 'GET'`.

Maybe, (better?, we rename `NetworkQuery` to `BaseNetworkQuery`; and apart, we create a union type `export type NetworkQuery = GetNetworkQuery | PostNetworkQuery | PutNetworkQuery | DeleteNetworkQuery;` which can be used in `NetworkDataSource.getRequestWithParameters` which I understand was the reason of exporting the base class.

File: packages/core/src/repository/query/network-query.ts
Comment (doup) [inline packages/core/src/repository/query/network-query.ts:6]: **Nit:** Is there a reason of using getter methods instead of regular methods? Any advantage over `getBody/getUrlParameters/getQueryParameters`?

**Review by doup — CHANGES_REQUESTED** [review]
Looks good in general, I've put "few" comments (🙏); and there are a couple of silly bugs. Maybe it would be nice to see some usage examples.

File: packages/core/src/repository/query/network-query.ts
Comment (urijp) [reply packages/core/src/repository/query/network-query.ts:6]: I could ask the same the other way around... These are "funtions" that return a structure based on the attributes, so no actions and no input parameters, so they are dynamic attributes.

**Comment by urijp** [comment]
No problem, I appreciate it, as commented, it's a WIP and needs a deeper revision and testing, but your comments are useful, even apart from that, I need to add more things, so I'll ask for another check next week.

File: packages/core/src/repository/query/network-query.ts
Comment (orioljp) [reply packages/core/src/repository/query/network-query.ts]: changed to path

File: packages/angular/src/data/http-request.builder.ts
Comment (orioljp) [reply packages/angular/src/data/http-request.builder.ts:83]: We have somehow the interface of our httpRequestBuilder coupled to the interface of the Datasources. Does it have any sense to set delete as <T | undefined>? Even if the api could return a body, I wouldn't set unknown everywhere.

**Comment by orioljp** [comment]
@lucianosantana @doup @joselufo @Alex-DA now you can check the whole PR, it's ready to be merged

**Review by doup — CHANGES_REQUESTED** [review]
Just two things (I've put them as comments also):

- `NetworkOneQuery` should expose `key` (?)
- I would like to see how query subclassing looks like as I've the impression that it can be improved, but I might be wrong.

File: packages/core/src/repository/query/network.query.ts
Comment (orioljp) [reply packages/core/src/repository/query/network.query.ts:37]: True! should be protected

File: packages/core/src/repository/query/network.query.ts
Comment (orioljp) [reply packages/core/src/repository/query/network.query.ts:11]: Let's make a suggestion @doup so we can discuss it. I have examples that I can show you but the current sample is not ready to apply this network datasource and we want to release this version ASAP.

Comment (doup) [reply packages/core/src/repository/query/network.query.ts:11]: As shared in Slack.
**Disclaimer:** I haven't tested these… I just checked that TS was happy.

```ts
interface NetworkQueryOptions {
    key?: string;
    method: HttpMethod;
    path?: string;
    body?: unknown;
    urlParameters?: Record<string, ParameterType>;
    queryParameters?: Record<string, ParameterType>;
}

export class NetworkQuery extends KeyQuery {
    private readonly data: NetworkQueryOptions;

    constructor(options: NetworkQueryOptions) {
        super(options.key ?? '');
        this.data = options;
    }

    get method(): HttpMethod {
        return this.data.method;
    }

    get path(): string {
        return this.data.path ?? '/';
    }

    get body(): unknown | undefined {
        return this.data.body;
    }

    get urlParameters(): Record<string, ParameterType> {
        return this.data.urlParameters ?? {};
    }

    get queryParameters(): Record<string, ParameterType> {
        return this.data.queryParameters ?? {};
    }
}

export class GetNetworkQuery extends NetworkQuery {
    constructor(options: Omit<NetworkQueryOptions, 'method'>) {
        super({
            ...options,
            method: HttpMethod.GET,
        });
    }
}

export class PostNetworkQuery extends NetworkQuery {
    constructor(options: Omit<NetworkQueryOptions, 'method'>) {
        super({
            ...options,
            method: HttpMethod.POST,
        });
    }
}

export class PutNetworkQuery extends NetworkQuery {
    constructor(options: Omit<NetworkQueryOptions, 'method'>) {
        super({
            ...options,
            method: HttpMethod.PUT,
        });
    }
}

export class DeleteNetworkQuery extends NetworkQuery {
    constructor(options: Omit<NetworkQueryOptions, 'method'>) {
        super({
            ...options,
            method: HttpMethod.DELETE,
        });
    }
}

interface NetworkOneQueryOptions extends Omit<NetworkQueryOptions, 'path' | 'urlParameters'> {
    pathOne: string;
    id: string | number;
}

export class NetworkOneQuery extends NetworkQuery {
    constructor(options: NetworkOneQueryOptions) {
        super({
            key: options.key,
            method: options.method,
            path: options.pathOne,
            urlParameters: { id: options.id },
            body: options.body,
            queryParameters: options.queryParameters,
        });
    }
}

export class GetOneNetworkQuery extends NetworkOneQuery {
    constructor(options: Omit<NetworkOneQueryOptions, 'method'>) {
        super({
            ...options,
            method: HttpMethod.GET,
        });
    }
}

export class PutOneNetworkQuery extends NetworkOneQuery {
    constructor(options: Omit<NetworkOneQueryOptions, 'method'>) {
        super({
            ...options,
            method: HttpMethod.PUT,
        });
    }
}

export class DeleteOneNetworkQuery extends NetworkOneQuery {
    constructor(options: Omit<NetworkOneQueryOptions, 'method'>) {
        super({
            ...options,
            method: HttpMethod.DELETE,
        });
    }
}
```

Examples:

```ts
class GetBudgetsQuery extends GetNetworkQuery {
    constructor(date: Date) {
        super({
            path: '/budgets',
            queryParameters: {
                date: date.toISOString(),
            },
        });
    }
}

class GetBudgetQuery extends GetOneNetworkQuery {
    constructor(id: number) {
        super({
            pathOne: '/budgets/:id',
            id,
        });
    }
}

class DeleteBudgetQuery extends DeleteOneNetworkQuery {
    constructor(id: number) {
        super({
            pathOne: '/budgets/:id',
            id,
        });
    }
}
```

File: packages/core/src/repository/query/network.query.ts
Comment (doup) [reply packages/core/src/repository/query/network.query.ts:37]: Well, I mean that it's not even in the constructor, not as a matter of visibility (private/protected).

File: packages/core/src/repository/query/network.query.ts
Comment (doup) [inline packages/core/src/repository/query/network.query.ts:12]: It's OK to default to `''`?? I understand we don't want to make this required. At the same time if we have the cache set and we forget to set the `key`… it can be confusing.

Does it make sense to create a random key instead? We have the `randomString` utility… This probably can also be confusing.

Maybe we should log a warning on the `CacheRepository` if the `key = ''`???

File: packages/core/src/repository/query/network.query.ts
Comment (doup) [inline packages/core/src/repository/query/network.query.ts:59]: We should also add a `PostOneNetworkQuery`, no?

File: packages/core/src/repository/query/network.query.ts
Comment (doup) [inline packages/core/src/repository/query/network.query.ts:3]: **Nit:** What do you think about adding a type alias for `Record<string, ParameterType>`?

```suggestion

export type NetworkParams = Record<string, ParameterType>;

```

File: packages/core/src/repository/query/network.query.ts
Comment (orioljp) [reply packages/core/src/repository/query/network.query.ts:59]: I don't think so, POST requests are made to create the entity so there's no id to complete the url, so it's not a valid use case.

File: packages/core/src/repository/query/network.query.ts
Comment (urijp) [reply packages/core/src/repository/query/network.query.ts:37]: solved!

File: packages/core/src/repository/query/network.query.ts
Comment (urijp) [reply packages/core/src/repository/query/network.query.ts:11]: I finally introduced the path through the constructor, but it's better to have body and queryParams in separate methods for clarity. The examples shared here are so simple and not considering big bodies, operations and so on, that would be in the constructor before the super (where we cannot call to any this.method).

**Review by doup — APPROVED** [review]

**Comment by doup** [comment]
@orioljp I'm approving it, if you do further changes and want another review tell me (or use the sidebar "Re-request review").

File: packages/core/src/repository/data-source/network.data-source.ts
Comment (orioljp) [reply packages/core/src/repository/data-source/network.data-source.ts:29]: You're right

**Pull request merged**
